### DISCUSSION TITLE
Downloading along a GPX track

### DISCUSSION THREAD

**Discussion by greuters** [streetcomplete/StreetComplete#5345] [Ideas]
Hello, thanks a lot for this great app @westnordost  and others!

We're currently on a long cycling trip and started adding data to OSM, especially the surface and quality of streets which is interesting to us (and often unclear here in Mexico). As we are riding on a tandem, we started to conveniently edit on the way and I hacked together a quick solution to download tiles along our planned route when we have WIFI in our overnight locations. See the (quick and dirty, but working) implementation here: [greuters/StreetComplete:gpx-download](https://github.com/streetcomplete/StreetComplete/compare/master...greuters:StreetComplete:gpx-download)
To download, there is a new button "Download data along gpx" next to the existing "Download here" button which lets you select a GPX track and starts downloading tiles to cover it.

![Screenshot_20231030_105002_StreetComplete Dev](https://github.com/streetcomplete/StreetComplete/assets/6537188/4949339e-2f65-4781-9019-8e59afe0fad4)
![Screenshot_20231030_105047_StreetComplete Dev](https://github.com/streetcomplete/StreetComplete/assets/6537188/b462c345-f9c9-444d-8545-f30ce37cfd43)

Do you think such a feature might be interesting to more users and be worth it to invest some more time in a proper solution, probably also overlaying the track on the map (only the last imported one)? I can imagine people going for a hike would use it, cyclists on tandem are probably too much of a niche.

If you like, I can clean up the code a bit and would of course be open for hints how to better do it. In related discussions (
#1757, #3898) I already saw that you think placing it in the hamburger menu is too prominent, would there be a more appropriate place?

I'll be available during the next two weeks, afterwards probably only with sporadic internet access, but there is no hurry for me.

Best,
simon

**Comment by westnordost** [comment]
Interesting! 

Be sure to use the (surface) overlay for such things, this is exactly the use case it was made for! :-)

Glad your solution works for you!  I previously outlined in this post https://github.com/streetcomplete/StreetComplete/issues/473#issuecomment-1777293054 how I imagine a final implementation could look like (I understand your implementation would fall short when the points in the GPX tracks are too far apart).

On the topic of whether this would be accepted as a PR, you are right that I previously stated that placing it in the hamburger menu would be too prominent... a more appropriate place... to be honest, none comes to mind, other than hiding it behind a TBD "expert mode" or something.

On that note, @Helium314 maintains a fork/extension to StreetComplete that basically intends to be pretty much that. And, it also already implements a "download along GPX", though, as far as I know, it only takes the bounding box of the whole GPX track, so this implementation could be enhanced with a better algorithm.

**Reply by Helium314** [reply]
see also SCEE issue https://github.com/Helium314/SCEE/issues/471

**Reply by westnordost** [reply]
> you are right that I previously stated that placing it in the hamburger menu would be too prominent... a more appropriate place... to be honest, none comes to mind, other than hiding it behind a TBD "expert mode" or something.

Thinking more about this, maybe a requirement for it to be included in mainline StreetComplete would be that you could actually draw the track along which you want to download within the app itself. In other words, there would be no import of a GPX, instead, a view would open (similar to the cut-way or move-node view) in which you could paint onto the map which cells should be downloaded. Could be fun. Does this make sense? Or would anyone planning a (bicycle) trip use a different app to generate a GPX anyway?

**Reply by westnordost** [reply]
Alternatively, maybe it is possible to register StreetComplete as an app that can open GPX. (Much like StreetComplete is registered to open Geo URIs.) So, when sharing a GPX, it can be shared with StreetComplete and the user has the option to download along the track then.

With this, the functionality would be there without requiring any additional buttons in the menu and thus, inclusing in mainline StreetComplete would be possible.

**Reply by Helium314** [reply]
> Alternatively, maybe it is possible to register StreetComplete as an app that can open GPX

For opening GPX files this should work with
```
            <intent-filter>
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data android:mimeType="application/gpx" android:scheme="content" android:host="*" android:pathPattern=".*\\.gpx" />
            </intent-filter>
```
Not sure about the mime type though, maybe should be `application/octet-stream`.

**Reply by greuters** [reply]
Nice, thank you for sharing your ideas!

I think registering an intent should be straightforward and I'll invest a little time in exploring that option. Planning in SC is a bit over the top I guess, also because good routing (e.g. BRouter) is included in many apps and probably most users already employ an app they are used to which is much more appropriate (better outdoor map styles, automatic routing, route editing, point management). I will also address the issue of interpolating intermediate waypoints if the track is too sparse (currently my solution would fallback to download the minimal bounding box including the two points - potentially way too much area).
=> this should give a minimal approach which can be considered for inclusion in SC, only major drawback being inefficient tile downloading, as each tile is downloaded separately.

Regarding integration in SCEE the major obstacle for me will be an efficient download strategy as mentioned in [SCEE #471](https://github.com/Helium314/SCEE/issues/471). I'll look into that but this will probably take me some time.

**Reply by mnalis** [reply]
I agree that it is best to stick to **StreetComplete supporting already existing `.gpx` files**; the intent idea is nice one, but not as important as being able to download along various types of gpx files. 

As for _"planning the route in SC"_, there is IMHO way too much that goes into that (searching for locations / amenities both internally and externally, different routing for cars and various types of bicycles, elevation data, traffic estimation, etc) to be able to put it into StreetComplete (on mobile, OsmAnd does allow for much but not all of that, and it is already hugely complex). Also even if StreetComplete did decide to become "route planner" (which I really don't think would be a good fit), the _external `.gpx` file support would still be needed_ (e.g. people acquire external .gpx files from other users who traveled there before, or from official websites who created routes, traveling agencies, QR codes etc).

**Comment by greuters** [comment]
A more mature implementation is available at [greuters:StreetComplete:gpx-download-by-intent](https://github.com/streetcomplete/StreetComplete/compare/master...greuters:StreetComplete:gpx-download-by-intent) now.

I tried to address most points from above, so this version:

- has no visible UI changes apart from two new dialogs which are opened when importing a track, and the track being shown on the map after import
- import is triggered by different intent-filters; I tested export / share from Locus Map and OsmAnd, as well as opening a GPX track from Samsung File Manager and Signal
- the algorithm to determine download bounding boxes is quite a bit more involved now; it guarantees a selectable sweep distance along the track is covered (interpolating track points if necessary) and merges adjacent bounding boxes if not "too much" additional area has to be downloaded because of the merge
- processing of the track (before scheduling downloads) can be aborted any time
- before scheduling downloads, the number of downloads and downloaded area are displayed and confirmation is asked
- downloads are scheduled via DownloadController, which handles caching quite well (importing the same track again after abort, overlapping downloads)
- if the import would download an area > 25 * ApplicationConstants.MAX_DOWNLOADABLE_AREA_IN_SQKM, it is automatically aborted and the user informed
 
The test tracks I played around with (100 - 200km) result in around 50 - 150 downloads between 50 - 200 sqkm and download volumes / times are still OK for me even in more populated areas (around 15min max e.g. around Zurich). What would you consider sane limits? Should the max area be restricted by some user setting?

I think the PR is ready for a first review and test, especially the UI part might have some pitfalls that I didn't see due to lack of Android experience..

**Reply by greuters** [reply]
Some screenshots of the current version:
![Import settings and file processing progress](https://github.com/streetcomplete/StreetComplete/assets/6537188/07e14fbc-1df0-4d3f-8c87-1569e6402ae3)
![Confirmation dialog](https://github.com/streetcomplete/StreetComplete/assets/6537188/c2ea9416-6c0b-4446-96bd-3572d7cbd9cf)
![Resulting download area](https://github.com/streetcomplete/StreetComplete/assets/6537188/abe4b9ca-5be2-4171-a9c8-29b22e36a361)

**Reply by mnalis** [reply]
> I think the PR is ready for a first review and test,

@greuters, I don't think you've actually opened [Pull Request](https://github.com/streetcomplete/StreetComplete/pulls) for it yet? 
Did you check if it works when gpx points are far apart as mentioned earlier, i.e.:

> Glad your solution works for you! I previously outlined in this post https://github.com/streetcomplete/StreetComplete/issues/473#issuecomment-1777293054 how I imagine a final implementation could look like (I understand your implementation would fall short when the points in the GPX tracks are too far apart).

Thanks for putting effort into this!

**Reply by greuters** [reply]
Yes, I didn't open the PR yet as I still want to add some unit tests. Having points far apart works, you can test it by setting the distance to download within low, e.g. to 10 meters. I tested down to 1m - which makes everything awfully slow ;-) - but nicely adds points in between the GPX waypoints.

**Reply by mnalis** [reply]
That's great news then! I didn't intend to rush you --  I'm not sure if you know you can open it as a `Draft PR` (by clicking on that triangle right of `Create pull request` button) which will indicate that PR is not ready for approvals / merging yet, but you want to allow people to comment and suggest code improvements on it.

**Reply by westnordost** [reply]
Wow, from a quick glance, this draft PR looks really well made! Actually, you can create a PR and mark it as draft if you are still working on it but have some questions or want to have feedback on some things.

96 downloads is massive, of course. I wouldn't be surprised if that hit some rate limits of the OSM API. If such a limit is hit, it should be handled gracefully - we do not want that it will result in an endless loop of retrying and retrying, for example. 

Now, for such huge downloads, it would be really useful if there was a cancel button (and/or pause/resume-button) in the download-notification but I guess this would be a general feature and not anything that is specifically related to this draft feature.

The display of the GPX track: How is it persisted? How to remove it again?

The first dialog is a bit odd, both from a UI perspective and from whether it makes sense to show it at all:
1. The checkbox seems misplaced: First you select in which distance you want to download, then you select whether you want to download at all? For the progress bar, I don't know what it is for... is it the progress for calculating which tiles to download? Does it take that long?
1. I dare say that the selection between 10 and 100 meters distance is probably almost moot, as the tiles in which it is downloaded will be somewhat larger anyway, no? Did you test what is the difference in terms  of elements downloaded and total time to download it it makes?

**Reply by westnordost** [reply]
I don't know for sure, but I would think that the number of downloads is substantial to driving up the overall download time and the time could be reduced by making less downloads but a larger area. I have some ideas for an algorithm, but you wrote you already implemented something like this, so I will shut my mouth until I get a proper look at your code.

**Reply by greuters** [reply]
Ok, thanks for the feedback. I opened it as draft PR now.

**UI**
- Rearranging the checkbox makes total sense, I'll do that. Maybe it would even make sense to hide / disable the selector when it is unchecked?
- The progress bar is there for processing the file and calculating which tiles to download. In DEV mode (connected with debugger) this process could take up to one minute, so I added progress indication. Might not be necessary any more now, on a decent phone from 2021 processing usually takes less than 10s with the current implementation and no debugger. Do you have any preference on this?

**download distance**
The distance determines the minimal area covered around the track, or in other words, how closely the track can sneak past a tile without having to download it (0 would mean to only download the tiles that are actually touched, though the current implementation would not support it). So even though tiles are (depending on your location, afaik) usually bigger, this makes for quite a difference.

Example results on a 130km test track:
- (tested with unlimited settings, 30s needed for processing the track) Download within 1m -> 148 downloads, 71km2
- Download within 10m -> 144 downloads, 75km2
- Download within 100m -> 100 downloads, 106km2
- Download within 250m -> 62 downloads, 185km2
- Download within 500m -> too big to handle, i.e. > 300km2
Download times scale with both number of downloads and area. Usually 100m was the best compromise for my tracks, though I did not do thorough testing on this.

**persistence**
This is an open question for me. At the moment, the track is simply lost when you close the app, or overwritten when you import another one. I think that might be ok, but if you would want it persisted this might be done similar to LocationAwareMapFragment.onSaveInstanceState. We might run into limits with this approach though, given your comment of max 5000 storable trackpoints.

**rate limits**
I have no clue how to handle that ATM - doesn't the DownloadController do this kind of stuff? I didn't care for yet ;)

**Reply by Helium314** [reply]
Did you check whether background map tiles are properly downloaded?
After doing many successive downloads, I sometimes notice missing tiles (usually only when it's too late and I'm at the place without internet connection). The cache is definitiely large enough, not hitting the limit.

**Reply by westnordost** [reply]
Link to PR #5369

Let's continue in that ticket.
